admin-api-server/src/nft/service/file/file.service.ts

```diff
@@ -0,0 +1,460 @@
+const __dirname = path.dirname(__filename);
+
+export interface File {
+  originalname: string;
+  fieldname: string;
+  mimetype: string;
+  encoding: string;
+  buffer: Buffer;
+  size: number;
+}
+
+type FileType = 'thumbnail' | 'display';
+type Path = string;
+type Dimension = {
+  width: number;
+  height: number;
+};
+
+interface ResizedImgParams {
+  buffer: Buffer;
+  imageSize: { width: number; height: number };
+  type: FileType;
+  extension: string;
+}
+
+interface ImageFromVideoParams {
+  videoFile: File;
+  typeOfImage: FileType;
+  imageSize: number;
+}
+
+interface GenerateImageFileParams {
+  source: Path | Buffer;
+  type: FileType;
+  extension: string;
+}
+
+interface ScreenshotFromVideoParams {
+  imageSize: number;
+  videoPath: string;
+  filename: string;
+}
+
+interface HandleVideoParams {
+  thumbnailMissing: boolean;
+  displayMissing: boolean;
+}
+
+interface HandleImageParams {
+  thumbnailMissing: boolean;
+  displayMissing: boolean;
+}
+
+@Injectable()
+export class FileService {
+  private MEDIA_TEMP_DIR: string;
+  private allFiles: File[] = [];
+  private artifact: File;
+
+  async addMissingFiles(filesArray: File[]): Promise<File[]> {
+    this.MEDIA_TEMP_DIR = __dirname + `/${uuidv4()}/`;
+    this.allFiles = filesArray;
+    this.artifact = filesArray.find(
+      (file) => file.originalname === 'artifact',
+    ) as File; // we always assume that we get an artifact from the client
+
+    const thumbnailMissing = !filesArray.some(
+      (file) => file.originalname === 'thumbnail',
+    );
+    const displayMissing = !filesArray.some(
+      (file) => file.originalname === 'display',
+    );
+    const isVideo = (this.artifact.mimetype.match(/video/g) || []).length > 0;
+    const isImage = (this.artifact.mimetype.match(/image/g) || []).length > 0;
+
+    try {
+      if (isVideo) {
+        this.#createTempDir();
+        const addedFiles = await this.#handleVideo({
+          thumbnailMissing,
+          displayMissing,
+        });
+
+        if (addedFiles.length) {
+          addedFiles.forEach((file) => {
+            this.allFiles.push(file);
+          });
+        }
+      } else if (isImage) {
+        this.#createTempDir();
+        const addedFiles = await this.#handleImage({
+          thumbnailMissing,
+          displayMissing,
+        });
+
+        if (addedFiles.length) {
+          addedFiles.forEach((file) => {
+            this.allFiles.push(file);
+          });
+        }
+      }
+    } catch (error) {
+      Logger.error(
+        `Failed during nft file imgae/video processing. err: ${error}`,
+      );
+    } finally {
+      if (fs.existsSync(this.MEDIA_TEMP_DIR)) {
+        this.#deleteTempDir();
+      }
+    }
+
+    return this.allFiles;
+  }
+
+  async #handleVideo({
+    displayMissing,
+    thumbnailMissing,
+  }: HandleVideoParams): Promise<File[]> {
+    let addedFiles: File[] = [];
+    if (displayMissing && thumbnailMissing) {
```

Comment:
This `if .. else if ..` block isn't great (as you already mentioned yourself). I don't think it's necessary if approached slightly differently:

1. start with display, if displayMissing then generate it.
2. if thumbnail is missing, generate it, take as source the generated display _or_ if this is missing then take the uploaded display (or reversed logic, also fine)

This will result in code something like this regarding execution branches:

```
if (displayMissing) {
  ...
}
if (thumbnailMissing) {
  display = displayMissing ? generatedDisplay : uploadedDisplay;
}
Easier to read and to maintain. Because we now no longer have essentially duplicated logic of generating the thumbnail in multiple execution branches.
```

Reply:
Definitely easier to read. Appreciate it. Should be better now.